diff --git a/arch/mips/boot/dts/ralink/mt7628a.dtsi b/arch/mips/boot/dts/ralink/mt7628a.dtsi
index 45bf96a3d..b47beacfa 100644
--- a/arch/mips/boot/dts/ralink/mt7628a.dtsi
+++ b/arch/mips/boot/dts/ralink/mt7628a.dtsi
@@ -215,6 +215,26 @@ i2c: i2c@900 {
 			status = "disabled";
 		};
 
+		i2s: i2s@a00 {
+                        compatible = "mediatek,mt7628-i2s";
+                        reg = <0xa00 0x100>;
+
+                        resets = <&resetc 17>;
+                        reset-names = "i2s";
+
+                        interrupt-parent = <&intc>;
+                        interrupts = <10>;
+
+                        txdma-req = <2>;
+                        rxdma-req = <3>;
+
+                        dmas = <&gdma 4>,
+                                <&gdma 6>;
+                        dma-names = "tx", "rx";
+
+                        status = "disabled";
+                };
+
 		uart0: uartlite@c00 {
 			compatible = "ns16550a";
 			reg = <0xc00 0x100>;
@@ -262,8 +282,57 @@ uart2: uart2@e00 {
 
 			reg-shift = <2>;
 		};
+
+		pwm: pwm@5000 {
+                        compatible = "mediatek,mt7628-pwm";
+                        reg = <0x5000 0x1000>;
+                        #pwm-cells = <2>;
+
+                        resets = <&resetc 31>;
+                        reset-names = "pwm";
+
+                        pinctrl-names = "default";
+                        pinctrl-0 = <&pinmux_pwm0_pwm>, <&pinmux_pwm1_pwm>;
+
+                        status = "disabled";
+                };
+
+		pcm: pcm@2000 {
+                        compatible = "ralink,mt7620a-pcm";
+                        reg = <0x2000 0x800>;
+
+                        resets = <&resetc 11>;
+                        reset-names = "pcm";
+
+                        interrupt-parent = <&intc>;
+                        interrupts = <4>;
+
+                        status = "disabled";
+                };
+
+                gdma: gdma@2800 {
+                        compatible = "ralink,rt3883-gdma";
+                        reg = <0x2800 0x800>;
+
+                        resets = <&resetc 14>;
+                        reset-names = "dma";
+
+                        interrupt-parent = <&intc>;
+                        interrupts = <7>;
+
+                        #dma-cells = <1>;
+                        #dma-channels = <16>;
+                        #dma-requests = <16>;
+
+                        status = "disabled";
+                };
 	};
 
+        clockc: clock-controller {
+                compatible = "ralink,rt2880-clock";
+                #clock-cells = <1>;
+        };
+
 	usb_phy: usb-phy@10120000 {
 		compatible = "mediatek,mt7628-usbphy";
 		reg = <0x10120000 0x1000>;
@@ -275,6 +344,19 @@ usb_phy: usb-phy@10120000 {
 		reset-names = "host", "device";
 	};
 
+        sdhci: sdhci@10130000 {
+                compatible = "ralink,mt7620-sdhci";
+                reg = <0x10130000 0x4000>;
+
+                interrupt-parent = <&intc>;
+                interrupts = <14>;
+
+                pinctrl-names = "default";
+                pinctrl-0 = <&pinmux_sdmode_sdxc>;
+
+                status = "disabled";
+        };
+
 	usb@101c0000 {
 		compatible = "generic-ehci";
 		reg = <0x101c0000 0x1000>;
@@ -286,6 +368,30 @@ usb@101c0000 {
 		interrupts = <18>;
 	};
 
+        ethernet: ethernet@10100000 {
+                compatible = "ralink,rt5350-eth";
+                reg = <0x10100000 0x10000>;
+
+                interrupt-parent = <&cpuintc>;
+                interrupts = <5>;
+
+                resets = <&resetc 21 &resetc 23>;
+                reset-names = "fe", "esw";
+
+                mediatek,switch = <&esw>;
+        };
+
+        esw: esw@10110000 {
+                compatible = "mediatek,mt7628-esw", "ralink,rt3050-esw";
+                reg = <0x10110000 0x8000>;
+
+                resets = <&resetc 23>;
+                reset-names = "esw";
+
+                interrupt-parent = <&intc>;
+                interrupts = <17>;
+        };
+
 	wmac: wmac@10300000 {
 		compatible = "mediatek,mt7628-wmac";
 		reg = <0x10300000 0x100000>;
diff --git a/arch/mips/boot/dts/ralink/vocore2.dts b/arch/mips/boot/dts/ralink/vocore2.dts
index fa8a5f8f2..01eb2f81e 100644
--- a/arch/mips/boot/dts/ralink/vocore2.dts
+++ b/arch/mips/boot/dts/ralink/vocore2.dts
@@ -15,4 +15,105 @@ chosen {
 		bootargs = "console=ttyS2,115200";
 		stdout-path = &uart2;
 	};
+
+	sound {
+                compatible = "simple-audio-card";
+                simple-audio-card,name = "soundcard";
+                simple-audio-card,format = "i2s";
+                simple-audio-card,bitclock-master = <&sound_master>;
+                simple-audio-card,frame-master = <&sound_master>;
+                simple-audio-card,mclk-fs = <272>;
+                simple-audio-card,widgets =
+                        "Microphone", "Microphone Jack",
+                        "Headphone", "Headphone Jack";
+                simple-audio-card,routing =
+                        "LINPUT1", "Microphone Jack",
+                        "RINPUT1", "Microphone Jack",
+                        "Microphone Jack", "Mic Bias",
+                        "Headphone Jack", "LOUT1",
+                        "Headphone Jack", "ROUT1";
+
+                simple-audio-card,cpu {
+                        sound-dai = <&i2s>;
+                };
+
+                sound_master:simple-audio-card,codec {
+                        sound-dai = <&codec>;
+                };
+        };
+};
+
+&watchdog {
+	status = "okay";
+};
+
+&spi {
+	status = "okay";
+
+	flash@0 {
+                compatible = "jedec,spi-nor";
+                reg = <0>;
+                spi-max-frequency = <10000000>;
+
+                partitions {
+                        compatible = "fixed-partitions";
+                        #address-cells = <1>;
+                        #size-cells = <1>;
+
+                        partition@0 {
+                                label = "u-boot";
+                                reg = <0x0 0x30000>;
+                                read-only;
+                        };
+
+                        partition@30000 {
+                                label = "u-boot-env";
+                                reg = <0x30000 0x10000>;
+                                read-only;
+                        };
+
+                        factory: partition@40000 {
+                                label = "factory";
+                                reg = <0x40000 0x10000>;
+                                read-only;
+                        };
+
+                        partition@50000 {
+                                compatible = "denx,uimage";
+                                label = "firmware";
+                                reg = <0x50000 0xfb0000>;
+                        };
+                };
+        };
+};
+
+&i2c {
+	status = "okay";
+
+	codec: es8388@10 {
+                #sound-dai-cells = <0>;
+                compatible = "everest,es8328", "evereset,es8388";
+                reg = <0x10>;
+
+                clocks = <&clockc 20>;
+                /* clock-names = "refclk"; */
+        };
+};
+
+&i2s {
+	#sound-dai-cells = <0>;
+	status = "okay";
+};
+
+&gdma {
+	status = "okay";
+};
+
+&pwm {
+	status = "okay";
+};
+
+&sdhci {
+	status = "okay";
+	mediatek,cd-poll;
 };
