diff --git a/package/kernel/linux/modules/other.mk b/package/kernel/linux/modules/other.mk
index 204fd0599f..ee851a8824 100644
--- a/package/kernel/linux/modules/other.mk
+++ b/package/kernel/linux/modules/other.mk
@@ -1289,3 +1289,11 @@ define KernelPackage/f71808e-wdt/description
 endef
 
 $(eval $(call KernelPackage,f71808e-wdt))
+
+define KernelPackage/vga-console
+  SUBMENU:=$(OTHER_MENU)
+  TITLE:=VGA text console
+  KCONFIG:= CONFIG_VGA_CONSOLE=y
+endef
+
+$(eval $(call KernelPackage,vga-console))
