Index: linux-5.15.153/drivers/gpu/drm/Kconfig
===================================================================
--- linux-5.15.153.orig/drivers/gpu/drm/Kconfig
+++ linux-5.15.153/drivers/gpu/drm/Kconfig
@@ -80,7 +80,7 @@ config DRM_DEBUG_SELFTEST
 	  If in doubt, say "N".
 
 config DRM_KMS_HELPER
-	tristate
+	tristate "KMS helper"
 	depends on DRM
 	help
 	  CRTC helpers for KMS drivers.
@@ -210,7 +210,7 @@ config DRM_KMS_CMA_HELPER
 	  Choose this if you need the KMS CMA helper functions
 
 config DRM_GEM_SHMEM_HELPER
-	bool
+	bool "SHMEM helper"
 	depends on DRM
 	help
 	  Choose this if you need the GEM shmem helper functions
