--- a/drivers/mmc/core/sdio.c
+++ b/drivers/mmc/core/sdio.c
@@ -1182,9 +1182,6 @@
 err:
 	mmc_detach_bus(host);
 
-	pr_err("%s: error %d whilst initialising SDIO card\n",
-		mmc_hostname(host), err);
-
 	return err;
 }
 
--- a/drivers/mmc/core/core.c
+++ b/drivers/mmc/core/core.c
@@ -1450,7 +1450,6 @@
 
 	ocr &= host->ocr_avail;
 	if (!ocr) {
-		dev_warn(mmc_dev(host), "no support for card's volts\n");
 		return 0;
 	}
 
