--- a/drivers/mmc/host/mtk-mmc/sd.c
+++ b/drivers/mmc/host/mtk-mmc/sd.c
@@ -440,6 +440,8 @@
 	else
 		inserted = (status & MSDC_PS_CDSTS) ? 1 : 0;
 
+	if (host->mmc->caps & MMC_CAP_NEEDS_POLL)
+		inserted = 1;
 #if 0
 	change = host->card_inserted ^ inserted;
 	host->card_inserted = inserted;
@@ -1857,6 +1859,9 @@
 			present = (sdr_read32(MSDC_PS) & MSDC_PS_CDSTS) ? 0 : 1;
 		else
 			present = (sdr_read32(MSDC_PS) & MSDC_PS_CDSTS) ? 1 : 0;
+
+		if (host->mmc->caps & MMC_CAP_NEEDS_POLL)
+			present = 1;
 		host->card_inserted = present;
 #endif
 		spin_unlock_irqrestore(&host->lock, flags);
