--- a/sound/soc/codecs/es8328.c
+++ b/sound/soc/codecs/es8328.c
@@ -39,6 +39,19 @@
 	.list	= rates_12288,
 };
 
+static const unsigned int rates_12000[] = {
+	8000, 12000, 16000, 24000, 32000, 44100, 48000, 96000,
+};
+
+static const int ratios_12000[] = {
+	27, 24, 23, 21, 20, 19, 18, 16,
+};
+
+static const struct snd_pcm_hw_constraint_list constraints_12000 = {
+	.count	= ARRAY_SIZE(rates_12000),
+	.list	= rates_12000,
+};
+
 static const unsigned int rates_11289[] = {
 	8018, 11025, 22050, 44100, 88200,
 };
@@ -552,6 +565,28 @@
 	return snd_soc_update_bits(codec, reg, ES8328_RATEMASK, ratio);
 }
 
+static int es8328_search_clock(struct snd_soc_codec *codec, int clk)
+{
+	int clocks[] = {12000000, 24000000, 11289600, 22579200, 12288000, 24576000};
+	int i;
+
+	for (i = 0; i < sizeof(clocks) / sizeof(int); i++) {
+		int cur = clk - clocks[i];
+		cur = cur > 0 ? cur : -cur;
+
+		/* if the clock is too far from request clock, skip. */
+		if (cur < clocks[i] / 1000)
+			break;
+	}
+
+	if (i == sizeof(clocks) / sizeof(int)) {
+		dev_err(codec->dev, "no fit clock find.\n");
+		return 0;
+	}
+
+	return clocks[i];
+}
+
 static int es8328_set_sysclk(struct snd_soc_dai *codec_dai,
 		int clk_id, unsigned int freq, int dir)
 {
@@ -559,11 +594,18 @@
 	struct es8328_priv *es8328 = snd_soc_codec_get_drvdata(codec);
 	int mclkdiv2 = 0;
 
-	switch (freq) {
+	switch (es8328_search_clock(codec, freq)) {
 	case 0:
 		es8328->sysclk_constraints = NULL;
 		es8328->mclk_ratios = NULL;
 		break;
+	case 24000000:
+		mclkdiv2 = 1;
+		/* fallthru */
+	case 12000000:
+		es8328->sysclk_constraints = &constraints_12000;
+		es8328->mclk_ratios = ratios_12000;
+		break;
 	case 22579200:
 		mclkdiv2 = 1;
 		/* fallthru */
@@ -784,26 +826,18 @@
 		return ret;
 	}
 
-	/* Setup clocks */
-	es8328->clk = devm_clk_get(codec->dev, NULL);
-	if (IS_ERR(es8328->clk)) {
-		dev_err(codec->dev, "codec clock missing or invalid\n");
-		ret = PTR_ERR(es8328->clk);
-		goto clk_fail;
-	}
-
-	ret = clk_prepare_enable(es8328->clk);
-	if (ret) {
-		dev_err(codec->dev, "unable to prepare codec clk\n");
-		goto clk_fail;
-	}
-
+	/* double speed mode is necessary. */
+	snd_soc_write(codec, ES8328_ADCCONTROL5, 0x20);
+	snd_soc_write(codec, ES8328_DACCONTROL2, 0x20);
+
+	/* set sound volumn to max for easy test. */
+	snd_soc_write(codec, ES8328_DACCONTROL17, 0x80);
+	snd_soc_write(codec, ES8328_DACCONTROL20, 0x80);
+	snd_soc_write(codec, ES8328_DACCONTROL4, 0x00);
+	snd_soc_write(codec, ES8328_DACCONTROL5, 0x00);
+	snd_soc_write(codec, ES8328_LOUT1VOL, 0x1e);
+	snd_soc_write(codec, ES8328_ROUT1VOL, 0x1e);
 	return 0;
-
-clk_fail:
-	regulator_bulk_disable(ARRAY_SIZE(es8328->supplies),
-			       es8328->supplies);
-	return ret;
 }
 
 static int es8328_remove(struct snd_soc_codec *codec)
