--- a/target/linux/ramips/modules.mk
+++ b/target/linux/ramips/modules.mk
@@ -136,3 +136,28 @@
 endef
 
 $(eval $(call KernelPackage,sound-mt7620))
+
+define KernelPackage/sound-mt7628
+  TITLE:=MT7628 PCM/I2S Alsa Driver
+  DEPENDS:=@TARGET_ramips +kmod-sound-soc-core +kmod-regmap-i2c +kmod-dma-ralink @!TARGET_ramips_rt288x
+  KCONFIG:= \
+        CONFIG_SND_RALINK_SOC_I2S \
+        CONFIG_SND_SIMPLE_CARD \
+        CONFIG_SND_SIMPLE_CARD_UTILS \
+        CONFIG_SND_SOC_ES8328 \
+	CONFIG_SND_SOC_ES8328_I2C
+  FILES:= \
+        $(LINUX_DIR)/sound/soc/ralink/snd-soc-ralink-i2s.ko \
+        $(LINUX_DIR)/sound/soc/generic/snd-soc-simple-card.ko \
+        $(LINUX_DIR)/sound/soc/generic/snd-soc-simple-card-utils.ko \
+        $(LINUX_DIR)/sound/soc/codecs/snd-soc-es8328.ko \
+	$(LINUX_DIR)/sound/soc/codecs/snd-soc-es8328-i2c.ko
+  AUTOLOAD:=$(call AutoLoad,90,snd-soc-es8328 snd-soc-es8328-i2c snd-soc-ralink-i2s snd-soc-simple-card)
+  $(call AddDepends/sound)
+endef
+
+define KernelPackage/sound-mt7628/description
+ Alsa modules for ralink i2s controller.
+endef
+
+$(eval $(call KernelPackage,sound-mt7628))
