diff --git a/main.c b/main.c
index 33cd734..9138682 100644
--- a/main.c
+++ b/main.c
@@ -29,8 +29,8 @@ int main(void)
     lv_disp_drv_init(&disp_drv);
     disp_drv.draw_buf   = &disp_buf;
     disp_drv.flush_cb   = fbdev_flush;
-    disp_drv.hor_res    = 800;
-    disp_drv.ver_res    = 480;
+    disp_drv.hor_res    = 480;
+    disp_drv.ver_res    = 800;
     lv_disp_drv_register(&disp_drv);
 
     evdev_init();
