diff --git a/lv_conf.h b/lv_conf.h
index 3137b1a..3f36997 100644
--- a/lv_conf.h
+++ b/lv_conf.h
@@ -30,7 +30,7 @@ extern uint32_t custom_tick_get(void);
  *====================*/
 
 /*Color depth: 1 (1 byte per pixel), 8 (RGB332), 16 (RGB565), 32 (ARGB8888)*/
-#define LV_COLOR_DEPTH 32
+#define LV_COLOR_DEPTH 16
 
 /*Swap the 2 bytes of RGB565 color. Useful if the display has an 8-bit interface (e.g. SPI)*/
 #define LV_COLOR_16_SWAP 0
